js/src/util/bn.ts

```diff
@@ -91,15 +91,25 @@ export function bigNumToNumber<T extends BN<BigNumArray>>(bn: T, scale?: number)
             number |= word * (BigInt(1) << BigInt(64 * i++));
         }
     }
-    if (typeof scale === 'number') {
-        const denominator = BigInt(Math.pow(10, scale));
+    if (typeof scale === 'number' && scale > 0) {
+        const denominator = BigInt('1' + '0'.repeat(scale));
         const quotient = number / denominator;
-        const remainder = number % denominator;
-        return bigIntToNumber(quotient) + (bigIntToNumber(remainder) / bigIntToNumber(denominator));
+        const remainder = negative? -(number % denominator) : number % denominator;
+        const integerPart = bigIntToNumber(quotient);
+        const fractionPart = padStart((remainder).toString(), scale);
+        const sign: string = negative && integerPart === 0 ? '-' : '';
+        return Number(`${sign}${integerPart}.${fractionPart}`);
```

Comment:
IIRC coercing via `+str` is a bit faster than using the Number constructor. Minifiers will typically do this as well:

```suggestion
        return +`${sign}${integerPart}.${fractionPart}`;
```